chatbot-ui/pages/index.tsx

```diff
@@ -25,4 +30,12 @@ function Home() {
   );
 }
 
-export default Home;
+export const getServerSideProps = (async () => {
+  return {
+    props: {
+      apiUrl:
+        process.env.NEXT_PUBLIC_CHAT_URL_API ??
```

Comment:
Just leaving a minor comment (hope you folks don't mind :) ) as I'm curious about this change. Just wondering why we're fetching public environment variables via server side props?  with Next.js you *should* be able to fetch public + server-side environment variables by using the getConfig function provided by next (as long as you've assigned this to publicRuntimeConfig/serverRuntimeConfig respectively in your next config file)

Reply:
Hi, no problem at all with you leaving a comment. Thanks for reviewing 🙂.

I think you've already figured most of this out but just thought I'd run through my understanding of it:

The `NEXT_PUBLIC_` variable is currently getting in-lined into the javascript during the Next standalone build that happens when building the docker image which ultimately ends up using `next build`.

In this change we are trying to read the API variable from the server environment at runtime instead of it being inlined at build time. That's for two reasons:

- We don't know the URL of the API until the API container is provisioned (at least while we have a temporary URL with no domain name set up). Assuming it is already provisioned I think we would then need to programmatically lookup the URL of the API and provide it to the environment for the `azd deploy` task which builds and deploys the container images using Docker so that it's available to the Next build which inlines it into the javascript. Alternatively we could write it to a `.env` file in the source so that it's available to the build to do the same thing. Both of these sound like a pain.

- We might want to later promote the same container image across different environments which won't work if the value is inlined in the javascript in the image.

It seems like the way to go is to set it as a server environment variable which can be picked up at runtime and it should be very easy to set this up in bicep for the web container to have a variable referring to the API container.

The advice on that page in the docs about using dynamic lookups was not working for me so I asked Amy to look into how to do this.

![image](https://github.com/dfe-analytical-services/chatbot-prototype/assets/4147126/35170076-45d9-44bb-ac32-49f63534020f)

You are right about the reason for us using `getServerSideProps`, although I'm not sure we'd specifically read the docs in enough detail to know this was the recommended approach or not. I'd defer to @amyb-hiveit on answering this!

I was aware that this was working for the EES public frontend (i.e. that it can read the server environment variables dynamically), but with not being too familiar with the frontend I wasn't sure how this had been achieved. Thanks for pointing it out.

I'm not sure if we should stick with the approach we've got here, or switch to the `getConfig` function that you mentioned. I can pick this up with @amyb-hiveit, but feel free to get in touch if you have an opinion.

I've renamed the variable from `NEXT_PUBLIC_CHAT_URL_API` to `CHAT_URL_API` to reflect that it's a server variable now.

I'm also not sure if this approach is entirely correct because I noticed if I put a `.env` file into the build it's value still takes precedence over the server environment.

To test it I added `CHAT_URL_API=http://env-file-var` to `chatbot-prototype/chatbot-ui/env.local`.

Then I built the image from the root of the project: `cd chatbot-prototype && docker build . -t chatbot-ui -f chatbot-ui/Dockerfile`.

Then I ran the container: `docker run --env CHAT_URL_API='http://server-var' -p 3000:3000 chatbot-ui:latest`.

The value it picks up with my env file existing in the source is `http://env-file-var` which goes against the docs which says `process.env` has the highest precedence 😕 .

When there's no env file it works fine though so I think I'm going to approve and merge this change.

Reading through the docs it sounds like migrating to App Router is the way to go for the future and that it would fix this problem. I'm not sure how much work this would be and given that this is a prototype I'm not sure we will be given a chance to prioritise it 😭 .

Reply:
Cool, yeh seems sensible to me. Moving to the app directory is not too necessary for 90% of uses cases tbh. Would make sense to move the main EES frontend over to it but it's quite fiddly to do in my experience (with regards to splitting out server + client components)